crates/precompile-evm-to-native-swap/src/tests.rs

```diff
@@ -0,0 +1,497 @@
+            ExitReason::Error(ExitError::Other("exactly one argument is expected".into())),
+            EvmDataWriter::new().build(),
+        );
+    });
+}
+
+/// This test verifies that the swap precompile call fails the call has an incomplete argument.
+#[test]
+fn swap_fail_short_argument() {
+    new_test_ext().execute_with_ext(|_| {
+        let expected_gas_usage: u64 = 50_000; // all passed gas will be consumed
+        let expected_fee: Balance = gas_to_fee(expected_gas_usage);
+
+        let mut input = EvmDataWriter::new_with_selector(Action::Swap).build();
+        input.extend_from_slice(&hex_literal::hex!("1000")); // bad input
+
+        run_failed_test_and_assert(
+            input,
+            U256::from(1),
+            expected_gas_usage,
+            expected_fee,
+            ExitReason::Error(ExitError::Other("exactly one argument is expected".into())),
+            EvmDataWriter::new().build(),
+        );
+    });
+}
+
+/// This test verifies that the swap precompile call fails when the call has extra data after
+/// the end of the first argument.
+#[test]
+fn swap_fail_trailing_junk() {
+    new_test_ext().execute_with_ext(|_| {
+        let expected_gas_usage: u64 = 50_000; // all passed gas will be consumed
+        let expected_fee: Balance = gas_to_fee(expected_gas_usage);
+
+        let mut input = EvmDataWriter::new_with_selector(Action::Swap)
+            .write(H256::from(target_swap_native_account().as_ref()))
+            .build();
+        input.extend_from_slice(&hex_literal::hex!("1000")); // bad input
+
+        run_failed_test_and_assert(
+            input,
+            U256::from(1),
+            expected_gas_usage,
+            expected_fee,
+            ExitReason::Error(ExitError::Other("junk at the end of input".into())),
+            EvmDataWriter::new().build(),
+        );
+    });
+}
+
+/// This test verifies that the swap precompile call fails when called without the sufficient balance.
+#[test]
+fn runner_fail_source_balance_no_funds() {
+    new_test_ext().execute_with_ext(|_| {
+        let storage_root = frame_support::storage_root(frame_support::sp_runtime::StateVersion::V1);
+
+        // Invoke the function under test.
+        let execerr = <Test as pallet_evm::Config>::Runner::call(
+            source_swap_evm_account(),
+            *PRECOMPILE_ADDRESS,
+            EvmDataWriter::new_with_selector(Action::Swap)
+                .write(H256::from(target_swap_native_account().as_ref()))
+                .build(),
+            U256::from(INIT_BALANCE + 1),
+            50_000, // a reasonable upper bound for tests
+            Some(*GAS_PRICE),
+            Some(*GAS_PRICE),
+            None,
+            Vec::new(),
+            true,
+            true,
+            None,
+            None,
+            <Test as pallet_evm::Config>::config(),
+        )
+        .unwrap_err();
+        assert!(matches!(execerr.error, pallet_evm::Error::BalanceLow));
+        assert_eq!(
```

Comment:
Maybe use an `assert_noop` or something?

Reply:
Thanks for noting, improved with `assert_noop` - [f785ce2](https://github.com/humanode-network/humanode/pull/1449/commits/f785ce2f818696c0c6a6f7a112be65ee92943849)